File: src/test/java/com/chimbori/crux/links/LinkUrlExtractorTest.java
package com.chimbori.crux.links;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.junit.Test;

import java.io.File;
import java.io.IOException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class LinkUrlExtractorTest {
  @Test
  public void testFindLink() {
    assertEquals("https://m.facebook.com/story.php?story_fbid=11111111111111111&id=1111111111&comment_id=11111111111111111&notif_t=comment_mention&notif_id=1111111111111111&ref=m_notif#11111111111111111",
        extractFromTestFile("https://m.facebook.com/notifications", "facebook_notification_single.html").findLink().linkUrl());
    assertEquals("https://hermit.chimbori.com/test-url",
        extractFromTestFile("https://hermit.chimbori.com", "image_extractor_simple_img.html").findLink().linkUrl());
    assertEquals("https://hermit.chimbori.com/test",
        extractFromTestFile("https://hermit.chimbori.com", "image_extractor_css_style.html").findLink().linkUrl());
  }

  private LinkUrlExtractor extractFromTestFile(String baseUrl, String testFile) {
    try {
      Document doc = Jsoup.parse(new File("test_data/" + testFile), "UTF-8");
      return LinkUrlExtractor.with(baseUrl, doc.body());
    } catch (IOException e) {
      fail(e.getMessage());
      return null;
    }
  }
}

File: src/test/kotlin/com/chimbori/crux/links/LinkUrlExtractorTest.kt
package com.chimbori.crux.links

import java.io.File
import java.io.IOException
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.jsoup.Jsoup
import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.junit.Test

class LinkUrlExtractorTest {
  @Test
  fun testFindLink() {
    assertEquals(
      "https://m.facebook.com/story.php?story_fbid=11111111111111111&id=1111111111&comment_id=11111111111111111&notif_t=comment_mention&notif_id=1111111111111111&ref=m_notif#11111111111111111".toHttpUrl(),
      extractFromTestFile(
        "https://m.facebook.com/notifications".toHttpUrl(),
        "facebook_notification_single.html"
      )?.findLink()?.linkUrl
    )
    assertEquals(
      "https://hermit.chimbori.com/test-url".toHttpUrl(),
      extractFromTestFile(
        "https://hermit.chimbori.com".toHttpUrl(),
        "image_extractor_simple_img.html"
      )?.findLink()?.linkUrl
    )
    assertEquals(
      "https://hermit.chimbori.com/test".toHttpUrl(),
      extractFromTestFile(
        "https://hermit.chimbori.com".toHttpUrl(),
        "image_extractor_css_style.html"
      )?.findLink()?.linkUrl
    )
  }

  private fun extractFromTestFile(baseUrl: HttpUrl, testFile: String) = try {
    val doc = Jsoup.parse(File("test_data/$testFile"), "UTF-8")
    LinkUrlExtractor(baseUrl, doc.body())
  } catch (e: IOException) {
    fail(e.message)
    null
  }
}
